### PR TITLE
[runtime] integrate mono 2018-02

### PR DESCRIPTION
**Pull request by lewurm**

### CODE REVIEW

File: mk/xamarin.mk [hunk 1]
```diff
@@ -1,6 +1,15 @@
 ifdef ENABLE_XAMARIN
-NEEDED_MACCORE_VERSION := 62c4393c94ed274acbc1edce259a3c22a461f922
-NEEDED_MACCORE_BRANCH := master
+
+ifneq ("$(wildcard $(topdir)/class/lib/$(PROFILE))","")
+-include $(MACCORE_PATH)/mk/versions.mk
+$(warning INCLUDED)
+@echo "   NEEDED_MONO_EXTENSIONS_VERSION=$(NEEDED_MONO_EXTENSIONS_VERSION)"
```
Comment (rolfbjarne) [inline mk/xamarin.mk]: This looks like debug spew.

File: mk/xamarin.mk [hunk 2]
```diff
@@ -58,6 +67,9 @@ test-$(1)::
 	@echo "   $(2)_PATH=$($(2)_PATH) => $(abspath $($(2)_PATH))"
 
 reset-$(1)::
+	@echo "   NEEDED_MONO_EXTENSIONS_VERSION=$(NEEDED_MONO_EXTENSIONS_VERSION)"
```
Comment (rolfbjarne) [inline mk/xamarin.mk]: Same: debug spew?

**Comment by luhenry** [comment]
This needs to be integrated in sync with XA. This **cannot** go in before XA is also ready to merge the `2018-02` integration

**Comment by lewurm** [comment]
@marek-safar there are two linker related failures left:
```
[FAIL] LinkSdkRegressionTest.UIButtonSubclass : VerifyIsUIButton 
[FAIL] LinkAllRegressionTest.Bug59015 : System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetNotificationForWaitCompletion 
```
are those buggy tests too?

**Comment by marek-safar** [comment]
@lewurm yep (should be fixed now)

**Comment by lewurm** [comment]
Hey @chamons, three tests in `CodeStrippingTests.cs` that you added recently ( https://github.com/xamarin/xamarin-macios/commit/ca4360116c24aa847f707dad9d7f1d2fb46834b4#diff-f62ae54ad2fad75dd18def0baf1d9d71R60 ) are failing:

```
Xamarin.MMP.Tests.CodeStrippingTests.ShouldStripMonoPosixHelper(null,False,True): Release lipo usage did not match expectations
Expected: True
But was: False

Xamarin.MMP.Tests.CodeStrippingTests.ShouldStripMonoPosixHelper(True,True,True): Release lipo usage did not match expectations
Expected: True
But was: False

Xamarin.MMP.Tests.CodeStrippingTests.ShouldStripMonoPosixHelper(False,False,False): System.IO.FileNotFoundException : Could not find file "/Users/builder/data/lanes/5929/a7abab1c/source/xamarin-macios/tests/xharness/tmp-test-dir/mmptest/bin/Debug/tmp-test-dir/Xamarin.MMP.Tests.MMPTests.RunMMPTest8/bin/Release/UnifiedExample.app/Contents/MonoBundle/libMonoPosixHelper.dylib"
```
any idea what could cause that with mono `2018-02`?

**Comment by lewurm** [comment]
Finally green 🎉  Reminder: the merge must be coordinated with XA https://github.com/xamarin/xamarin-android/pull/1289

**Comment by marek-safar** [comment]
@spouliot are you ok with merging this and on what date?

**Comment by lewurm** [comment]
@marek-safar afaik we wait for the okay by QA, tracked here: https://devdiv.visualstudio.com/DevDiv/Xamarin%20QA/_workitems/edit/587386

(also it's broken right now, should be green again soon)

**Comment by spouliot** [comment]
@marek-safar sure, as soon as we have a working build and it's gone thru one of the bots with devices

File: tests/linker/ios/link all/LinkAllTest.cs [hunk 3]
```diff
@@ -551,13 +551,14 @@ public void WebKit_NSProxy ()
...
+		static Type type_Task = typeof (Task);
+
 		[Test]
 		public void Bug59015 ()
 		{
 			CheckAsyncTaskMethodBuilder (typeof (AsyncTaskMethodBuilder));
 			CheckAsyncTaskMethodBuilder (typeof (AsyncTaskMethodBuilder<int>));
-			var t = typeof (Task);
-			var snfwc = t.GetMethod ("SetNotificationForWaitCompletion", BindingFlags.Instance | BindingFlags.NonPublic);
+			var snfwc = type_Task.GetMethod ("SetNotificationForWaitCompletion", BindingFlags.Instance | BindingFlags.NonPublic);
```
Comment (spouliot) [inline tests/linker/ios/link all/LinkAllTest.cs]: @marek-safar can you explain why those changes were _needed_ ?

File: tests/linker/ios/link all/InterfacesTest.cs [hunk 4]
```diff
@@ -78,18 +82,18 @@ public void Bug10866 ()
 			F (new A ());
 
 			// Foo and Bar methods are both used on A and must be present
-			Assert.NotNull (typeof(A).GetMethod ("Foo", BindingFlags.Instance | BindingFlags.Public), "A::Foo");
-			Assert.NotNull (typeof(A).GetMethod ("Bar", BindingFlags.Instance | BindingFlags.Public), "A::Bar");
+			Assert.NotNull (type_a.GetMethod ("Foo", BindingFlags.Instance | BindingFlags.Public), "A::Foo");
+			Assert.NotNull (type_a.GetMethod ("Bar", BindingFlags.Instance | BindingFlags.Public), "A::Bar");
```
Comment (spouliot) [inline tests/linker/ios/link all/InterfacesTest.cs]: Has something changed in the linker so the old and new code is not seen as identical ?
That smells like a breaking change, we can update the test not the existing (customer) code.

Comment (marek-safar) [reply tests/linker/ios/link all/InterfacesTest.cs]: The linker is now smarter so the test old code which used this approached to check for non-existence of the method actually forced the linker not to remove it.

**Comment by lewurm** [comment]
@spouliot 
> [...] and it's gone thru one of the bots with devices

on what does that depend? VSTS step on wrench says
> No VSTS URL found, the build was probably not scheduled successfully.VSTS build for inspector status: completed (canceled)

**Comment by rolfbjarne** [comment]
@lewurm there seems to be two random System test failures on device:

> [FAIL] HttpResponseStreamTest.AutomaticDecompression : System.Net.WebException : Error getting response stream (ReadAsync): ReceiveFailure Unable to read data from the transport connection: Connection reset by peer.
> [FAIL] SocketTest.TestSelect1 : #02 

otherwise the device tests look good.

There is also breaking API in the BCL:

screenshot 2018-04-06 10 32 16

**Comment by lewurm** [comment]
Thanks @rolfbjarne, this aligns with what QA reported: https://github.com/xamarin/xamarin-macios/issues/3827

@marek-safar @EgorBo  is the API diff expected?

**Comment by marek-safar** [comment]
@rolfbjarne this was already discussed 2 months ago

GetData is expected to be missing because it was Mono bug, the rest is bug/limitation of api-diff not to include base type members

**Comment by marek-safar** [comment]
@spouliot are we still waiting for any confirmation here?

**Comment by lewurm** [comment]
@rolfbjarne I can't reproduce it (I ran my tests on an iPad Air and iPhone7). It can't be network specific too because the test is doing things on the loopback device:
https://github.com/mono/mono/blob/42d425cb62db513146b1a463e22323f4d5b577f4/mcs/class/System/Test/System.Net.Sockets/SocketTest.cs#L253-L315

What is suspicious though, in the report for the device tests http://xamarin-storage/macios-mac-mono-2018-02/95/950f2ebb502fdbf13389b6533fe7785a826a17aa/ae9c43ecc00f/jenkins-results/tests/index.html , I see that the failing tests were executed on `XQAiPadPro1`; on the other arm64 device `XQAiPhone6Sd` the failure didn't happen. Maybe something device specific? On the other hand, I checked device builds on master and didn't see anything like that. So, either:
- it was a bad luck run
- it's indeed related to `2018-02` and hard to reproduce.

Let's wait for the next device results 🤔

**Comment by lewurm** [comment]
it happened again on `XQAiPadPro1`: http://xamarin-storage/macios-mac-mono-2018-02/c2/c2701e159c1401b8bc9166c92ff5ec998fcf66db/71bfec61173a/jenkins-results/tests/index.html  

     
screen shot 2018-04-19 at 12 27 01

**Comment by filipnavara** [comment]
Well, `SocketTest.TestSelect1` looks a bit sketchy to me in regards to the expected thread switching interaction. Maybe, just maybe, there is something wrong in the test itself.

**Comment by lewurm** [comment]
tried to reproduce the tvOS simulator failure locally, no luck. hmm 😕

**Comment by lewurm** [comment]
seeing this on the mono master lane too:
https://jenkins.mono-project.com/job/test-mono-pull-request-i386-osx/9798/testReport/junit/(root)/SocketTest/TestSelect1/

**Comment by rolfbjarne** [comment]
There's this failure on the device bots too:

> [FAIL] HttpClientTest.Proxy_Disabled : needs internet access 

but it's random, and most likely a network problem on the devices, and since this is the only other problem (not also in master) you can consider the device tests green.

**Pull request merged**